specification/dartLangSpec.tex

```diff
@@ -11267,33 +11267,24 @@ \subsection{Function Expressions}
 which is used below and in other sections, as follows:
 
 \begin{itemize}
-\item If $T$ is \code{FutureOr<$S$>} for some $S$ then $\flatten{T} = S$.
+\item If $T$ is a type variable $X$ with bound $B$,
+  or $T$ is a promoted type variable $X & B$,
+  then $\flatten{T} = \flatten{B}$.
 
-\item Otherwise if
-  \code{$T <:$ Future}
-  then let $S$ be a type such that
-  \code{$T <:$ Future<$S$>}
-  and for all $R$, if
-  \code{$T <:$ Future<$R$>}
-  then $S <: R$.
+\item Otherwise if $T$ is \code{FutureOr<$S$>} for some $S$
```

Comment:
Here's an updated version of the rest of the definition of flatten (replacing lines 11274--11280, including both):

```
\item If $T$ is \code{$S$?}\ for some $S$
  then \DefEquals{\flatten{$T$}}{\code{\flatten{$S$}?}}.
\item If $T$ is \code{FutureOr<$S$>} then \DefEquals{\flatten{$T$}}{S}.
\item If $T$ implements \code{Future<$S$>}
  (\ref{interfaceSuperinterfaces})
  then \DefEquals{\flatten{$T$}}{S}.
\item Otherwise,
  \DefEquals{\flatten{$T$}}{T}.
```

Reply:
We'd need to add the following to `dart.sty`, (preferably, just before `% ......` at the end, such that it will merge cleanly with the dart.sty from the null safety update):

```
\newcommand{\DefEquals}[2]{\ensuremath{{#1}\stackrel{\vartriangle}{=}{#2}}}
\newcommand{\DefEqualsNewline}[2]{
  \ensuremath{{#1}\stackrel{\vartriangle}{=}}\\
  \ensuremath{{#2}}%
}
```